### ISSUE TITLE
PickRandom needs predicting

### ISSUE THREAD

**Issue by metalgearsloth** [space-wizards/space-station-14#15897]
Title.

**Comment by deltanedas** [comment]
the only way to predict would be doing the random on client right?

**Comment by ElectroJr** [comment]
In general, without just special casing it for this specific interaction, it would probably require a proper networked/synced random system with different seeds/sources of random data, but I have no idea how to best design that. AFAIK you'd probably:

- Need to reset the RNG state when the client resets predicted entities
- Make sure the state is occasionally sent to players to prevent desync
- Need a state per player (or entity?)
- Ensure that the order in which random numbers are generated is the same.
  - AFAIK there isn't even any kind of guarantee that client & server systems update in the same order.
  - Same goes for anything that enumerates over entities.

The first 2 could be fixed by just having a component with the seed & letting the game state code handle it, though then you'd have to reset the seed every tick and it seems generally inefficient.

The last point is the most complicated and I have no idea if there is some general way of tackling it. I guess you could just have each source of numbers be very specialized so that ordering issues should never be a problem? So either a source of randomness per interaction/system or entity.

See also space-wizards/RobustToolbox/issues/2857

**Comment by 0x6273** [comment]
I don't think this should be predicted, since what gets picked is supposed to be unpredictable. Shared RNG state wouldn't work here since clients would then know beforehand what is going to be picked.

What should be predicted here is adding the verb. All verbs should be predicted and "waiting for server" should be killed since it causes the menu to jump around making it easy to missclick.

**Comment by mirrorcult** [comment]
even if we had synced client/server rng this just should by its nature not be predictable so clients cant figure out what they're getting next

**Comment by ElectroJr** [comment]
> even if we had synced client/server rng this just should by its nature not be predictable so clients cant figure out what they're getting next

For important things like role selection or whatever: sure, but this will probably only be used for card games and the like where laggy interactions would feel shit and the stakes aren't high. In this case the benefit of adding cheat-prevention is just outweighed by the downside. If it ever gets used for anything important, you could just disable netsync and remove the client-side PickRandom / random-syncing component on that entity.

**Comment by 0x6273** [comment]
If you really want to predict this, then maybe you could have a placeholder item that gets spawned? For example for a deck of cards, you pick random and get a blank card, so there's at least some immediate feedback, and then it turns into the actual card you picked when the client gets it from the server.

**Comment by SlamBamActionman** [comment]
Will be resolved with this RT PR I think https://github.com/space-wizards/RobustToolbox/pull/5849
